### DISCUSSION TITLE
Does the Claw Mech actually work right?

### DISCUSSION THREAD

**Discussion by Moburma** [swfans/syndwarsfx#177] [General]
The claw mech things are oddly quite neglected in the game, they only really appear on the penultimate level, despite being a cool design. They're really just tanks in disguise, but it's a good disguise.
The thing is though... they don't really look right. 

![](https://i.ibb.co/K2sWwLN/Desktop-Screenshot-2024-06-08-17-04-10-12.png)
(Startframe 30 on the left, the default 29 on the right. I should have got a picture of the back of the red one, you can clearly see it's the intended design there, with a grey coloured grill that is barely visible on the black one).

Look at them, they are all black and have no detail on them at all. Looks very weird - it's not showing the textures properly, is it?. This weirdness increases 100 fold when you consider that there is a _much_ more complete and detailed design for the same thing. In the final game, the claw mech uses StartFrame 29 to be placed in the game. However, using StartFrame 30 uses the much better looking red variant, with what are clearly actual proper textures. 
So just use StartFrame 30, right? But no, you can't do that, it's not a complete and working variant, it's missing the second tank legs Thing (tanks are two Things joined together, tracks/legs and turret/body) and is just a body hovering in space and doesn't work right.

The plot thickens even further when you look at the PlayStation version of the game. This only uses the second red variant throughout the game, the weird looking black one is not seen at all. A less convincing further claim - using the old [Syndicate Vehicles Viewer ](http://syndicate.lubiki.pl/downloads/bullfrog_utils_swars_vehicles_reader.zip)tool there is no such thing as the black mech design shown at all, only the red model, suggesting the final one is some kind of mistake/bug.

Cutting to the chase, my hypothesis is this:

- The claw mech is supposed to be red
- The final game black design is not actually an intended variant, but some kind of bug/unfinished version that never got resolved before the game was forced out the door

I was looking in the code but didn't really understand how StartFrame influences the vehicle Things (I know it _does_ change the vehicle model, but I couldn't see _how_ it does it). Looking closely it does have the correct texture, it's just in the wrong colours - possibly it's using the wrong palette or applying some kind of incorrect effect on the model?

Thoughts?

**Comment by mefistotelis** [comment]
Maybe it has max shade value set by mistake. Or maybe it's on purpose and it was supposed to bring stealth fighter vibe.
Hard to tell.
The 3D objects redrawing is not remade.
(redrawing the "child" thing of a tank is done by `draw_rot_object()` function)

**Comment by Moburma** [comment]
Another weird thing that I think I am correct on - Claw mechs can only face one way. 

All other vehicles can be setup to face a particular way (in 3 dimensions in fact, even looking up and down for placement on hills), by setting the extended values on their Thing data to transform their model. This doesn't seem to work with claw mechs, they always only face one direction in a level and seem to ignore what is set here. 
In fact I think you can see this in the screenshot above - everyone is looking down towards the shuttle, but the mechs are facing the wrong way, looking away from it. If you look at the one on the left I replaced, this actually DOES seem to be using what is set in the level file, it's set at a 45 degree angle in line with part of the "arrow" formed by the line of zealots. I imagine both mechs are supposed to be facing inwards towards the shuttle like this.

This is also visible on the multiplayer level C065L005.DAT - this has loads of mechs and they are all facing the same way

**Comment by geist22** [comment]
Interesting find. I actually never looked close enough at them to notice the texture beneath the black. Seeing that darkened texture does indeed look weird.

But I don't think that your restored version is the intended one either. The texture is used for the syndicate's APCs and has a more "military" look, to me this doesn't look like something the Church would use, while the darkened version has a more "mysterious" feel to it that is more in line with the Church's theme.

Just my thoughts. But if the PlayStation version has the textured one, it probably is intended like this. Are there any screenshots of the PS versions anywhere?

**Reply by Moburma** [reply]
I'm 99% sure it is supposed to look that way now. It actually is in the Zealot colour scheme, red body and white tail pipes/legs. See the (slightly altered) PlayStation version below:

![](https://i.ibb.co/bBLhqW6/SLUS-002-62-23062024-191149-0815.png)
![](https://i.ibb.co/MSy1pFF/SLUS-002-62-23062024-191243-0920.png)

It's a little different here as it has yellow and black chevrons on the claws - I think the shiny metallic texture effect is not actually in the PS1 version at all. What is interesting here is that they still haven't fixed the positioning bug on the mechs, they're supposed to be facing inwards at a 45 degree angle. I guess by this point no one even realised it was broken.

**Comment by mefistotelis** [comment]
I wonder what the original idea for the claw tanks attack was.

The mech has all the limbs separate, to allow for animation without skeletal movement and even without mesh deformation.
But the claws are not separate.

Was the idea that it will just sweep the area with the claws, making it short distance tank? Or maybe the claws were supposed to be extendable on a string?

**Reply by Moburma** [reply]
Ahh wow, I didn't realise that, I definitely think there was more going on with them than meets the eye and we just got whatever they salvaged at the last minute. Perhaps as you say the claws would extend out and crush things?

**Reply by mefistotelis** [reply]
> Perhaps as you say the claws would extend out and crush things?

Or it would do the "come over here!" move, and then do short distance attack. Though when I'm looking at the claws I'm not so sure whether this was supposed to be short distance..

**Comment by Moburma** [comment]
Belated realisation - the main tank has this same visual bug too.
![](https://i.ibb.co/52FL1Tn/Desktop-Screenshot-2024-10-22-22-54-58-07.png)

Look at it from the back - why are the missile pods all black? Why is the back of the turret all black? They're again not like this in the PS1 version. It's one of those things I always assumed was meant to be this way, but looking more closely it appears to be wrong.

I wonder if it's because they are both "extra" models used by this special vehicle type and it was just never implemented properly. The shading problem is more noticeable on the claw mechs because the main chassis of the tank draws correctly, but 90% of the mech is the main body/turret part which is visibly wrong. 

I also need to check if the main tank turret behaves in the same way in how it faces on level start - my hypothesis is that the bottom model draws correctly facing the right way (tanks definitely do), but I think it's the top model that doesn't get placed correctly on level start. Unfortunately for the claw mechs that basically IS the vehicle so it's noticeably weird. If am right, then all tanks in a level should have their _turrets_ facing the same way on level start.

**Reply by Moburma** [reply]
Looks like I was right.

![](https://i.ibb.co/cTS66Sd/Desktop-Screenshot-2024-10-23-11-04-05-30.png)

Two tanks, one facing 90 degrees away from the other, but both have their turrets facing to the same place on level start. So I guess that is why the claw mechs all face the same way - with tank vehicles their "turret" is not oriented the same way as the base. I guess it's the simple fact that these vehicles consist of two Things on top of each other, and there is no orientation data for the second Thing, so it always just points to 0.

**Reply by Moburma** [reply]
Another tank observation that's a somewhat weaker point: If you hold down the mouse button to keep firing missiles from the tanks, they constantly flash dark and light. I assume this is supposed to be a kind of muzzle flash effect, simulating the light of the missile engine lighting up the tank as it leaves. However, what's weird is that it keeps constantly flashing every half second or so, when the tank missile fire rate is of course much slower than that. 

Bit of a reach here, but we know the tank was originally a twin barrelled projectile design (think of the mammoth tanks from C&C Red Alert ) - maybe this was timed to be in sync with the cannons firing, which then makes it look odd with the completely different final missile battery design?

What makes this much less likely an explanation is that it also uses the same effect when the tank is taking damage, so possibly it was just supposed to be a bullet impact type effect and they just recycled it for the missiles launching too, as most people aren't going to be holding the button after launch so wouldn't notice it's "wrong".

**Comment by geist22** [comment]
The flashing happens with all vehicles when someone inside acquires a target and is trying to attack, like when you're running around a tank faster than the turret can turn and I think also when the shooter inside a car is out of energy. Though a tank that is actively shooting doesn't flash (with an NPC inside).

I've been wondering how tanks work for a while, in particular the unguided vs. homing missiles. We know that the optimal way to have non-active agents use the Launcher is to equip them with Brain Lv3, which makes them fully charge chargeable weapons. But the tank's rockets aren't charged and fire instantly once the turret is fully turned towards the target. I know for a fact that the enemy tanks in EuroCorp mission 17 (Cairo) and 18 (Sevastopol) have fired homing missiles at my agents in the past, because I needed to retreat with the agent that was targeted by the tank, while the other 3 agents attacked it. I've been replaying those and other missions trying to reproduce it, but for some reason they only shot unguided missiles during my tests.

What is also interesting is that the range at which the tank operators acquire targets is dependant on their active weapon rather than the range of the rockets. Though once aggro'd, they keep firing even outside the rockets' range. So the ideal play would be to equip tank operators with the LR Rifle.

I still haven't figured out what makes enemies shoot guided or unguided missiles, or how you might manipulate your agents to always shoot guided missiles with a tank.

**Comment by mefistotelis** [comment]
The dark tanks are caused by this chunk of code in some drawing functions:
```
        ushort mag;
        mag = (p_face4->WalkHeader + (p_face4->GFlags >> 2)) & 7;
        if (mag != 0)
        {
            point1.S = (mag * point1.S) >> 3;
            point2.S = (mag * point2.S) >> 3;
            point3.S = (mag * point3.S) >> 3;
            point4.S = (mag * point4.S) >> 3;
        }
```

It seem to dampen the Shiness/Shade value based on some surprising parameters. I have no idea what was the intention behind that.

**Reply by Moburma** [reply]
I assume this function is used by other stuff? What I'm getting at is if it's just accidentally triggering it, or specifically designed to look this way.

**Reply by mefistotelis** [reply]
> What I'm getting at is if it's just accidentally triggering it, or specifically designed to look this way.

`¯\_(ツ)_/¯`

**Reply by mefistotelis** [reply]
There are two parameters used to dampen the lightness: `WalkHeader` and `GFlags`.
* `WalkHeader` is used to figure out from which face you can walk on another face when walking over 3D objects. But you cannot walk on vehicles, right? So, is it unset, or does it have a different meaning for faces making up a vehicle? Or maybe it is set like everywhere else, you just never walk on vehicles because there's no way to "jump" on it.
* `GFlags` are various flags, mostly used for walking as well. The darkening is determined by 3 of them: 0x04, 0x08 and 0x10. They seem to control whether you can walk on the face from somewhere else.

So, why are faces darkened based on walk parameters? Are these really walk parameters, or are the values reused for something else?
No idea.

**Reply by Moburma** [reply]
Coming back to this, I guess the weird thing is why does it only process like this on the _second_ Thing of the mech? If you look at the picture at the start of the thread, the model on the left is when you use the "wrong" mech Thing, this draws what would normally be the second model (the turret) alone, and in this case it renders it in a way that looks normal to me. It seems based on that it's only processing the model in this dark way when it's drawing both Things. Another example, here is the actual tank turret Thing on its own:
 ![](https://i.ibb.co/mtvRsvn/Desktop-Screenshot-2024-11-08-20-51-00-31.png)
So there are definitely instances where it draws it correctly, it just seems in the way it is drawn in the final use case it's "wrong" (I'm assuming it is wrong). So from my annoying layman logic I would say it's probably something to do with that, i.e. they got two Things drawing on top of each other in a basic way as required, but never went back and refined the bugs?

**Reply by mefistotelis** [reply]
Yeah, that looks like the base is updated to be drawn correctly, and the "child" thing - the turret - is not treated the same way and is not updated.

That would mean the solution will be to sync some properties (copy them from base to turret).

The code I write does the same thing as Bullfrog code, but it is different in structure, names, macros and details. So I suspect in the original code it could have been harder to figure out why the tanks got dark. And the timelines for releasing a game are always tight.

**Comment by Moburma** [comment]
Some shameless off topic created because this reminded me you can actually make [pretty much any object a vehicle](https://www.youtube.com/watch?v=L0jhRiaNKWw)

**Comment by mefistotelis** [comment]
Finally figured it out:

The darkening is an effect which should be applied at edge of the map - the vehicle gets darker before it disappears at map border.

So yeah, it shouldn't apply to tanks which are not at map border. It applies for the same reason you could hear sounds at edge of the map - because the turret has a relative position, which means the position is a small number. If interpreted as absolute position, it would be near edge - so darkening is applied.

**Reply by Moburma** [reply]
Ahh, excellent work, makes sense. One thing I wondered was that in the single picture of the old double cannon tank in previews it visibly wasn't dark, but then they probably just hadn't put the effect in yet. 

EDIT:  Doesn't that look better!

![](https://i.ibb.co/0FC6wbw/Desktop-Screenshot-2024-11-14-16-58-27-55.png)

The other cool thing would be if the turret inherited the facing angle from the bottom tank model on placement in the level. As you can see from the original screenshot the mechs are supposed to be facing towards the shuttle at 45 degree angles to it.

**Reply by mefistotelis** [reply]
> if the turret inherited the facing angle from the bottom tank model on placement in the level

Yeah, makes sense. You can create an "issue" for that.

**Comment by geist22** [comment]
Do you think the turrets of "neutral" tanks (those that haven't attacked yet - after acquiring a target, the turret will keep pointing at that point forever) should turn with the tank? The turret's direction is now aligned with how the lower part spawns at the start of the level, but when the tank does rotate, the turret keeps pointing in the same direction and not rotating with it to keep pointing "forward".

**Reply by mefistotelis** [reply]
When I first saw that the tank moves independently of the turret, I was actually impressed - the real tanks do that as well. So I wouldn't want to remove that completely.

There could be a correction after some delay, to make sure the player sees that the tank can move and turn without losing the target.
But is it worth implementing? not sure.
